syntax = "proto3";

option csharp_namespace = "Api.Grpc";

package geocode;

// Сервис геокодирования
service GeocodeService {
  // Поиск адреса по нормализованной строке
  rpc SearchAddress (SearchAddressRequest) returns (SearchAddressResponse);

  // Health check для мониторинга
  rpc HealthCheck (HealthCheckRequest) returns (HealthCheckResponse);
}

// ============================================================================
// Поиск адреса
// ============================================================================

// Запрос на поиск адреса
message SearchAddressRequest {
  // Нормализованная строка поиска (обработана через Address Parser)
  string normalized_query = 1;

  // Максимальное количество результатов (опционально, по умолчанию 10)
  int32 limit = 2;

  // ID сессии/запроса для трекинга и логирования
  string request_id = 3;

  // Дополнительные параметры поиска (опционально)
  SearchOptions options = 4;

  // Оригинальный запрос пользователя (до нормализации)
  string original_query = 5;

  // Структурированные компоненты адреса (результат ParseAddress из libpostal)
  ParsedAddressComponents parsed_components = 6;
}

// Опции поиска
message SearchOptions {
  // Минимальный порог релевантности (0-1)
  double min_score_threshold = 1;

  // Включить fuzzy search (нечеткий поиск)
  bool enable_fuzzy_search = 2;

  // Фильтр по населенному пункту (опционально)
  string locality_filter = 3;
}

// Ответ с результатами поиска
message SearchAddressResponse {
  // Статус выполнения запроса
  ResponseStatus status = 1;

  // Строка, по которой производился поиск
  string searched_address = 2;

  // Найденные объекты
  repeated AddressObject objects = 3;

  // Общее количество найденных результатов (может быть больше чем objects.length)
  int32 total_found = 4;

  // Метаданные запроса
  ResponseMetadata metadata = 5;
}

// Найденный адресный объект
message AddressObject {
  // Наименование населенного пункта (город, село, и т.д.)
  string locality = 1;

  // Наименование улицы
  string street = 2;

  // Номер дома (может содержать корпус, строение: "10к2с1")
  string number = 3;

  // Долгота (longitude)
  double lon = 4;

  // Широта (latitude)
  double lat = 5;

  // Оценка релевантности результата (0-1, где 1 - точное совпадение)
  double score = 6;

  // Дополнительная информация (опционально)
  AdditionalInfo additional_info = 7;
}

// Дополнительная информация об объекте
message AdditionalInfo {
  // Почтовый индекс
  string postal_code = 1;

  // Район города
  string district = 2;

  // Полный отформатированный адрес
  string full_address = 3;

  // Уникальный идентификатор объекта в БД
  string object_id = 4;
}

// Структурированные компоненты адреса (из libpostal)
message ParsedAddressComponents {
  // Номер дома
  string house_number = 1;

  // Улица
  string road = 2;

  // Квартира/офис
  string unit = 3;

  // Этаж
  string level = 4;

  // Подъезд
  string staircase = 5;

  // Вход
  string entrance = 6;

  // Почтовый индекс
  string postcode = 7;

  // Район/микрорайон
  string suburb = 8;

  // Город
  string city = 9;

  // Округ города
  string city_district = 10;

  // Округ
  string county = 11;

  // Регион/область
  string state = 12;

  // Округ региона
  string state_district = 13;

  // Страна
  string country = 14;

  // Регион страны
  string country_region = 15;

  // Остров
  string island = 16;

  // Мировой регион
  string world_region = 17;

  // Ближайшая точка интереса
  string near = 18;
}

// ============================================================================
// Общие типы
// ============================================================================

// Статус выполнения запроса
message ResponseStatus {
  // Код статуса
  StatusCode code = 1;

  // Сообщение (для ошибок или предупреждений)
  string message = 2;

  // Детали ошибки (опционально, для отладки)
  string details = 3;
}

// Коды статуса
enum StatusCode {
  OK = 0;                          // Успешное выполнение
  INVALID_REQUEST = 1;             // Некорректный запрос
  NOT_FOUND = 2;                   // Ничего не найдено
  INTERNAL_ERROR = 3;              // Внутренняя ошибка сервера
  TIMEOUT = 4;                     // Превышено время ожидания
  DATABASE_ERROR = 5;              // Ошибка базы данных
  CANCELLED = 6;                   // Запрос отменен
  SERVICE_UNAVAILABLE = 7;         // Сервис недоступен
}

// Метаданные ответа
message ResponseMetadata {
  // Время выполнения запроса в миллисекундах
  int64 execution_time_ms = 1;

  // Timestamp обработки запроса
  int64 timestamp = 2;

  // Версия поискового движка
  string engine_version = 3;
}

// ============================================================================
// Health Check
// ============================================================================

message HealthCheckRequest {
  // Пустой запрос
}

message HealthCheckResponse {
  // Статус здоровья сервиса
  HealthStatus status = 1;

  // Версия сервиса
  string version = 2;

  // Время работы сервиса в секундах
  int64 uptime_seconds = 3;
}

enum HealthStatus {
  HEALTHY = 0;
  DEGRADED = 1;
  UNHEALTHY = 2;
}
